### PR TITLE
TypeTransformer for reading and writing from TensorFlowRecord format

### PR DESCRIPTION
**Pull request by ryankarlos**
# TL;DR
This flyte feature adds support for users to read and write from `.tfrecord` file formats
using Tensorflow Example as a native type.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

* Adds a `TensorflowExampleTransformer` type  in `flytekit/extras/tensorflow/records.py` which uses the [tf.train.Example] (https://www.tensorflow.org/api_docs/python/tf/train/Example) message, and then serialize, write, and read tf.train.Example messages to and from `.tfrecord` files, following the examples in the Tensorflow docs https://www.tensorflow.org/tutorials/load_data/tfrecord
* Adds tests for serialisation and deserialisation steps in Transformer `tests/flytekit/unit/extras/tensorflow/test_transformations.py`
* Adds test for example workflow using tf.train.Example message.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2571

### CODE REVIEW

**Comment by ryankarlos** [comment]
I havent added this as a plugin since the original issue description was to add this feature similar to format of pytorch transformer type

**Comment by dennisobrien** [comment]
The unit test failures seem to be caused by `tensorflow` not being included.

> E   ModuleNotFoundError: No module named 'tensorflow'

You should be able to add this to `dev-requirements.in`.

I'm excited to see more tensorflow support being contributed!

File: dev-requirements.in [hunk 1]
```diff
@@ -13,3 +13,4 @@ google-cloud-bigquery
 google-cloud-bigquery-storage
 IPython
 torch
+tensorflow<=2.8.1
```
Comment (ryankarlos) [inline dev-requirements.in:16]: Ive had to pin the versions of tensorflow here as the latter ones would throw a error when trying to resolve protobuf dependencies - i guess  due to pinned protobuf version in requirements.txt (also i rather not make any changes to requirements.in to resolve this) 

Screenshot 2022-10-20 at 18 36 35

**Comment by ryankarlos** [comment]
> The unit test failures seem to be caused by `tensorflow` not being included.
> 
> > E   ModuleNotFoundError: No module named 'tensorflow'
> 
> You should be able to add this to `dev-requirements.in`.
> 
> I'm excited to see more tensorflow support being contributed!

@dennisobrien  thanks, i pushed the changes now. Ive also created a PR https://github.com/flyteorg/flytekit/pull/1242 for keras model support !

File: dev-requirements.in (hunk 1 above)
Comment (ryankarlos) [reply dev-requirements.in]: Ignore this - ive pinned `grpcio-status<1.49.0` instead based on suggestion from @pingsutw  in another PR

Comment (dennisobrien) [reply dev-requirements.in]: I recently created this PR https://github.com/flyteorg/flytekit/pull/1248 that adds version constraints to `grpcio` and `grpcio-status` in `requirements.in`.  You should be able to pull that change in now that it has been merged and avoid the constraint in `dev-requirements.in`.

Bug description here: https://github.com/flyteorg/flyte/issues/3006

Comment (ryankarlos) [reply dev-requirements.in]: I still get the same issue as above if pulling in from master . `grpcio` and `grpcio-status`  versions in requirements.in are:

```
grpcio<=1.47.0
grpcio-status<=1.47.0
```

I think  the it may need to be pinned `grpcio-status<1.49.0` as  @pingsutw had suggested (at least that was working for me) but not sure.

File: flytekit/extras/tensorflow/records.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import pathlib
+from typing import Type, TypeVar
```
Comment (pingsutw) [inline flytekit/extras/tensorflow/records.py:2]: Thank you @ryankarlos. Overall LGTM, could you put these files to `flytekit/extras/tensorflow/record/`

**Comment by ryankarlos** [comment]
@pingsutw pushed requested changes

**Comment by samhita-alla** [comment]
@cosmicBboy, thanks for writing this up! I like the idea behind `TFRecordFile` and `TFRecordsDirectory`. The directory format might be more useful, but I think we also need to support storing a single `tf.Train.Example` or `tf.data.Dataset`. 

Concerning your questions:

* I agree; we don't need a `tf.Train.Example` TypeTransformer
* I think so
* Am I right in assuming that `dataset` here corresponds to `TFRecordFile` or `TFRecordsDirectory`? If so, besides kwargs, we might also need to let users call [methods](https://www.tensorflow.org/api_docs/python/tf/data/TFRecordDataset#methods_2), e.g., see how [`get_dataset` fetches the data](https://keras.io/examples/keras_recipes/creating_tfrecords/#train-a-simple-model-using-the-generated-tfrecords) from a `TFRecordDataset`. But I don't think it's possible to streamline this into a type; so a better alternative will be to enable users to provide kwargs and let them apply additional methods or parsers if needed within a task, and I think this could facilitate extraction of the data from a `TFRecordDataset` to some extent.

As for the code structure, will this go into `flytekit/extras` directory?

**Comment by cosmicBboy** [comment]
> But I don't think it's possible to streamline this into a type; so a better alternative will be to enable users to provide kwargs and let them apply additional methods or parsers if needed within a task

Right, I'm thinking for the `tf.data.TFRecordDataset` annotated type, we'd just handle the initialization of the object `tf.data.TFRecordDataset(filenames, **kwargs)` and then pass that into the task, the user is responsible for other transformations in the function body:

```python
@task
def consume_records(
    dataset: Annotated[
        tf.data.TFRecordDataset,
        # configure kwargs to the constructor
        # https://www.tensorflow.org/api_docs/python/tf/data/TFRecordDataset
        TFRecordDatasetConfig(...)
    ]
):
    dataset = (
        dataset
        .map(parse_tfrecord_fn, num_parallel_calls=AUTOTUNE)
        .map(prepare_sample, num_parallel_calls=AUTOTUNE)
        .shuffle(batch_size * 10)
        .batch(batch_size)
        .prefetch(AUTOTUNE)
    )
```

What do you think? If this looks good I can update the proposal

> As for the code structure, will this go into flytekit/extras directory?

Yep! As long as we follow the same conventions as the `pytorch` extra I think we should make this part of the main flytekit api.

**Comment by samhita-alla** [comment]
@cosmicBboy looks good to me! @ryankarlos please read through the comments.

**Comment by ryankarlos** [comment]
Im getting this error when running `test_native.py` even though i have added the `tensorflow.python.data.ops.readers.TFRecordDatasetV2` type in `TensorfloTensorflowRecordsTransformer`. Any ideas ?

```
==================================== ERRORS ====================================
_ ERROR collecting tests/flytekit/unit/extras/tensorflow/records/test_native.py _
tests/flytekit/unit/extras/tensorflow/records/test_native.py:45: in <module>
    def wf():
flytekit/core/workflow.py:739: in workflow
    return wrapper(_workflow_function)
flytekit/core/workflow.py:734: in wrapper
    workflow_instance.compile()
flytekit/core/workflow.py:614: in compile
    workflow_outputs = exception_scopes.user_entry_point(self._workflow_function)(**input_kwargs)
flytekit/exceptions/scopes.py:198: in user_entry_point
    return wrapped(*args, **kwargs)
tests/flytekit/unit/extras/tensorflow/records/test_native.py:48: in wf
    consume(dataset=tf.data.TFRecordDataset(filenames=[generate_tf_record_file()], name="t1"))
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/data/ops/readers.py:458: in __init__
    filenames = _create_or_validate_filenames_dataset(filenames, name=name)
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/data/ops/readers.py:66: in _create_or_validate_filenames_dataset
    filenames = ops.convert_to_tensor(filenames, dtype_hint=dtypes.string)
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/profiler/trace.py:183: in wrapped
    return func(*args, **kwargs)
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/framework/ops.py:1640: in convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:343: in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:267: in constant
    return _constant_impl(value, dtype, shape, name, verify_shape=False,
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:279: in _constant_impl
    return _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:304: in _constant_eager_impl
    t = convert_to_eager_tensor(value, ctx, dtype)
../../opt/miniconda3/envs/flyte/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py:102: in convert_to_eager_tensor
    return ops.EagerTensor(value, ctx.device_name, dtype)
E   ValueError: Attempt to convert a value (Promise(node:n1.o0)) with an unsupported type (<class 'flytekit.core.promise.Promise'>) to a Tensor.


------------------------------- Captured stderr --------------------------------
{"asctime": "2022-11-04 21:42:05,906", "name": "flytekit", "levelname": "WARNING", "message": "Unsupported Type <class 'tensorflow.python.data.ops.readers.TFRecordDatasetV2'> found, Flyte will default to use PickleFile as the transport. Pickle can only be used to send objects between the exact same version of Python, and we strongly recommend to use python type that flyte support."}
2022-11-04 21:42:05.914428: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.

```

**Comment by dennisobrien** [comment]
Within the `workflow` function, everything is a `Promise`.  Can you try doing the work within a task?

```python
def wf():
    file = generate_tf_record_file()
    tf_record_dataset = get_tf_record_dataset_from_file(file)
    consume(dataset=tf_record_dataset)

```

File: docs/source/extras.tensorflow.rst [hunk 3]
```diff
@@ -0,0 +1,7 @@
+############
+Tensorflow Type
```
Comment (samhita-alla) [inline docs/source/extras.tensorflow.rst:2]:
```suggestion
TensorFlow Type
```

File: flytekit/extras/tensorflow/__init__.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+"""
+Flytekit Tensorflow
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:2]:
```suggestion
Flytekit TensorFlow
```

File: flytekit/extras/tensorflow/__init__.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+=========================================
+.. currentmodule:: flytekit.extras.tensorflow
+
+.. autosummary::
+   :template: custom.rst
+   :toctree: generated/
+
+    TensorflowRecord
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:10]:
```suggestion
    TensorFlowRecord
```

File: flytekit/extras/tensorflow/__init__.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    # isolate the exception to the tensorflow import
+    import tensorflow
+
+    _tensorflow_installed = True
+except (ImportError, OSError):
+    _tensorflow_installed = False
+
+
+if _tensorflow_installed:
+    from .records import TensorflowExampleRecordsTransformer
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:26]:
```suggestion
    from .record import TensorFlowRecordTransformer
```
Is it okay to change `records.py` to `record.py`?

File: flytekit/extras/tensorflow/__init__.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+else:
+    logger.info("We won't register TensorflowRecordsTransformer because tensorflow is not installed.")
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:28]:
```suggestion
    logger.info("We won't register TensorFlowRecordTransformer because tensorflow is not installed.")
```

File: flytekit/extras/tensorflow/records.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+T = TypeVar("T")
+
+
+@dataclass_json
+@dataclass
+class TFRecordDatasetConfig:
+    """
+    TFRecordDatasetConfig is the config used during the creating tf.data.TFRecordDataset comprising
+    record of one or more TFRecord files.
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/records.py:27]:
```suggestion
    TFRecordDatasetConfig can be used while creating tf.data.TFRecordDataset comprising
    one or more TFRecord files.
```

File: flytekit/extras/tensorflow/records.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    Args:
+      compression_type:A scalar evaluating to one of "" (no compression), "ZLIB", or "GZIP".
+      buffer_size:the number of bytes in the read buffer.If None, a sensible default for both local and remote file systems is used.
+      num_parallel_reads:the number of files to read in parallel. If greater than one, the records of files read in parallel are outputted in an interleaved order.
+      name:A name for the operation.
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/records.py:33]:
```suggestion
      compression_type: A scalar evaluating to one of "" (no compression), "ZLIB", or "GZIP".
      buffer_size: The number of bytes in the read buffer. If None, a sensible default for both local and remote file systems is used.
      num_parallel_reads: The number of files to read in parallel. If greater than one, the records of files read in parallel are outputted in an interleaved order.
      name: A name for the operation.
```

File: flytekit/extras/tensorflow/records.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+
+def extract_metadata(t: Type[T]) -> Tuple[T, TFRecordDatasetConfig]:
+    metadata = TFRecordDatasetConfig()
+    if get_origin(t) is Annotated:
+        base_type, metadata = get_args(t)
+        if isinstance(metadata, TFRecordDatasetConfig):
+            return base_type, metadata
+        else:
+            raise TypeTransformerFailedError(f"{t}'s metadata needs to be of type Dict")
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/records.py:49]:
```suggestion
            raise TypeTransformerFailedError(f"{t}'s metadata needs to be of type TFRecordDatasetConfig")
```

File: flytekit/extras/tensorflow/records.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    return t, metadata
+
+
+class TensorflowRecordsTransformerBase(TypeTransformer, Generic[T]):
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/records.py:53]:
```suggestion
class TensorFlowRecordTransformerBase(TypeTransformer, Generic[T]):
```

File: flytekit/types/directory/__init__.py [hunk 12]
```diff
@@ -26,3 +27,12 @@
     This is usually the SummaryWriter output in PyTorch or Keras callbacks which record the history readable by
     TensorBoard.
 """
+
+
+tfrecords_dir = typing.TypeVar("tfrecords_dir")
+TFRecordsDirectory = FlyteDirectory[tfrecords_dir]
```
Comment (samhita-alla) [inline flytekit/types/directory/__init__.py:33]: This format should be modified (see https://github.com/jerempy/flytekit/blob/bcab554eb69e31ca1844bf5c5a06a45b401e45f2/flytekit/types/file/__init__.py PR for reference). You can merge the latest master branch to fetch and merge the updates.

File: flytekit/extras/tensorflow/records.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            literal_type.blob is not None
+            and literal_type.blob.dimensionality == _core_types.BlobType.BlobDimensionality.SINGLE
+            and literal_type.blob.format == self.TENSORFLOW_FORMAT
+        ):
+            return T
+
+        raise ValueError(f"Transformer {self} cannot reverse {literal_type}")
+
+
+class TensorflowExampleRecordsTransformer(TensorflowRecordsTransformerBase[example_pb2.Example]):
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/records.py:126]: I'm guessing the type here needs to be `TFRecordFile` and `TFRecordsDirectory`. You haven't established a link between the flytefile and flytedirectory you defined with this transformer. Also, I'm not sure how not to associate two transformers with `TFRecordFile` — a FlyteFile transformer and this. I'm guessing it just needs to be a `typing.NewType`, not a FlyteFile type. WDYT @cosmicBboy?

Comment (samhita-alla) [reply flytekit/extras/tensorflow/records.py:126]: Do we also want to associate `tf.data.Dataset` with `TFRecordFile` and `TFRecordsDirectory`?

File: flytekit/extras/tensorflow/records.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
+        meta = BlobMetadata(
+            type=_core_types.BlobType(
+                format=self.TENSORFLOW_FORMAT,
+                dimensionality=_core_types.BlobType.BlobDimensionality.SINGLE,
+            )
+        )
+        local_dir = ctx.file_access.get_random_local_directory()
+        remote_path = ctx.file_access.get_random_remote_directory()
+        if isinstance(python_val, TFRecordsDirectory):
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/records.py:76]: Shouldn't the `python_val` be `tf.train.Example`?

Comment (samhita-alla) [reply flytekit/extras/tensorflow/records.py:76]: I think the `TypeTransformer` has to be such that it subclasses both `TFRecordFile` and `TFRecordsDirectory`. In the `to_literal` method, it needs to validate whether the type of `python_val` is `tf.train.Example` or a list of `tf.train.Example`s and thereby, return a blob with [dimensionality set to single or multipart](https://github.com/flyteorg/flytekit/blob/9e4069058845f018f410703ec7c4e3971a9ee85a/flytekit/models/core/types.py#L37-L39).

File: dev-requirements.txt [hunk 15]
```diff
@@ -134,8 +141,16 @@ google-api-core[grpc]==2.10.2
 google-auth==2.14.1
     # via
     #   google-api-core
+    #   google-auth-oauthlib
     #   google-cloud-core
+<<<<<<< HEAD
```
Comment (samhita-alla) [inline dev-requirements.txt:146]: Could you remove these?

File: flytekit/extras/tensorflow/__init__.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    # isolate the exception to the tensorflow import
+    import tensorflow
+
+    _tensorflow_installed = True
+except (ImportError, OSError):
+    _tensorflow_installed = False
+
+
+if _tensorflow_installed:
+    from .record import TensorflowRecordFileTransformer, TensorflowRecordsDirTransformer
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:26]: Could you change the names of classes from `Tensorflow...` to `TensorFlow...`?

File: flytekit/extras/tensorflow/__init__.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+else:
+    logger.info("We won't register TensorflowRecordTransformer because tensorflow is not installed.")
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:28]:
```suggestion
    logger.info("We won't register TensorFlowRecordFileTransformer and TensorFlowRecordsDirTransformer because tensorflow is not installed.")
```

File: flytekit/extras/tensorflow/records.py (hunk 13 above)
Comment (cosmicBboy) [reply flytekit/extras/tensorflow/records.py]: I'm a little confused by the implementation of [this spec](https://github.com/flyteorg/flytekit/pull/1240#issuecomment-1292623521)...

I think `TFRecordFile` and `TFRecordsDirectory` should follow the conventions for [flyte files](https://github.com/flyteorg/flytekit/blob/master/flytekit/types/file/__init__.py#L108-L111) and [directories](https://github.com/flyteorg/flytekit/blob/master/flytekit/types/directory/__init__.py#L22), then implement a separate `TFRecordsFileTransformer` and `TFRecordsDirectoryTransformer`, which subclass `FlyteFile` and `FlyteDirectory`, respectively, overriding the `to_python` and `to_literal` methods as necessary to implement the tensorflow-specific implementation.

I wouldn't worry to much about DRY'ing this code by implementing a single transformer for all three of these types... I think a transformer for each of these types will be fine and actually easier to maintain.

Shared functionality can be implemented as helper functions so we don't conflate transformers in a single transformer class implementation.

File: flytekit/extras/tensorflow/record.py [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    if get_origin(t) is Annotated:
+        base_type, metadata = get_args(t)
+        if isinstance(metadata, TFRecordDatasetConfig):
+            return base_type, metadata
+        else:
+            raise TypeTransformerFailedError(f"{t}'s metadata needs to be of type TFRecordDatasetConfig")
+    return t, metadata
+
+
+class TensorflowRecordTransformerBase(TypeTransformer, Generic[T]):
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/record.py:53]: Can we have two separate transformers for the file and directory? That would facilitate the maintenance of code.
cc: @cosmicBboy

File: flytekit/types/directory/__init__.py [hunk 19]
```diff
@@ -11,10 +11,15 @@
...
    TensorboardLogs
+   TFRecordsDirectory
 """
 
 import typing
 
+from typing_extensions import Annotated
+
+from flytekit.types.file import tfrecords_file
+
```
Comment (samhita-alla) [inline flytekit/types/directory/__init__.py:22]: Seems like this code block isn't required.

File: flytekit/types/file/__init__.py [hunk 20]
```diff
@@ -19,6 +19,8 @@
    PythonPickledFile
    PythonNotebook
    SVGImageFile
+   TFRecordFile
+   TFRecordsDirectory
```
Comment (samhita-alla) [inline flytekit/types/file/__init__.py:23]: This isn't required.

File: flytekit/extras/tensorflow/record.py (hunk 18 above)
Comment (samhita-alla) [reply flytekit/extras/tensorflow/record.py:53]: You can, however, include helper functions in case they need to be used in both transformers.

File: tests/flytekit/unit/extras/tensorflow/records/test_native.py [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+
+@task
+def generate_tf_record_dir() -> TFRecordsDirectory:
+    features = tf.train.Features(feature=dict(a=a, b=b, c=c))
+    return [tf.train.Example(features=features)] * 2
+
+
+@task
+def consume(dataset: Annotated[tf.data.TFRecordDataset, TFRecordDatasetConfig(name="testing")]):
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/records/test_native.py:39]: Can we also have one such task for `TFRecordFile` with the metadata?

File: tests/flytekit/unit/extras/tensorflow/records/test_native.py [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+@workflow
+def wf():
+    file = generate_tf_record_file()
+    files = generate_tf_record_dir()
+    consume(dataset=file)
+    consume(dataset=files)
+
+
+@workflow
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/records/test_native.py:52]: `@workflow` isn't required.

File: tests/flytekit/unit/extras/tensorflow/records/test_native.py (hunk 21 above)
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/records/test_native.py:39]: Can we also have a test with more complex config, please?

File: tests/flytekit/unit/extras/tensorflow/records/test_transformations.py [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+    example = tf.train.Example(features=features).SerializeToString()
+    with tf.io.TFRecordWriter(path) as writer:
+        for i in range(2):
+            writer.write(example)
+
+
+@pytest.mark.parametrize(
+    "transformer,python_type,format",
+    [
+        (TensorflowRecordFileTransformer(), TFRecordFile, TensorflowRecordFileTransformer.TENSORFLOW_FORMAT),
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/records/test_transformations.py:44]: Can you also add directory to this list? Also seems like you haven't used the right transformer name.

File: tests/flytekit/unit/extras/tensorflow/records/test_native.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
+from typing import Annotated
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/records/test_native.py:1]: Could you change the directory name from `records` to `record` and move `test_native.py` to `test_record.py`?

**Comment by ryankarlos** [comment]
Screenshot 2022-11-29 at 01 38 46

**Comment by ryankarlos** [comment]
Test failures on CI are unrelated to tests in this PR

Screenshot 2022-11-29 at 01 47 04

**Comment by samhita-alla** [comment]
Can you import `Annotated` from `typing_extensions`? That should fix the failures.

File: flytekit/extras/tensorflow/record.py [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+
+from flytekit.core.context_manager import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer, TypeTransformerFailedError
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Literal, Scalar
+from flytekit.models.types import LiteralType
+from flytekit.types.directory import TFRecordsDirectory
+from flytekit.types.file import TFRecordFile
+
+T = TypeVar("T")
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/record.py:18]: Are we using this anywhere?

File: flytekit/extras/tensorflow/record.py [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+    if get_origin(t) is Annotated:
+        _, metadata = get_args(t)
+        if isinstance(metadata, TFRecordDatasetConfig):
+            return uri, metadata
+        else:
+            raise TypeTransformerFailedError(f"{t}'s metadata needs to be of type TFRecordDatasetConfig")
+    return uri, metadata
+
+
+def to_tf_record_dataset_from_dir(
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/record.py:68]: Can we merge this and the file functions with the `to_python_val` methods? Seems like these functions aren't being re-used anywhere else, right? So I think it's okay to have the code within the transformers.

File: flytekit/extras/tensorflow/record.py [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+class TensorFlowRecordFileTransformer(TypeTransformer[TFRecordFile]):
+    """
+    TypeTransformer that supports serialising and deserialising to and from TFRecord file.
+    https://www.tensorflow.org/tutorials/load_data/tfrecord
+    """
+
+    TENSORFLOW_FORMAT = "TensorFlowRecord"
+
+    def __init__(self):
+        super().__init__(name="TensorFlow Record Directory", t=TFRecordFile)
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/record.py:110]:
```suggestion
        super().__init__(name="TensorFlow Record File", t=TFRecordFile)
```

File: flytekit/extras/tensorflow/record.py [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+class TensorFlowRecordsDirTransformer(TypeTransformer[TFRecordsDirectory]):
+    """
+    TypeTransformer that supports serialising and deserialising to and from TFRecord directory.
+    https://www.tensorflow.org/tutorials/load_data/tfrecord
+    """
+
+    TENSORFLOW_FORMAT = "TensorFlowRecord"
+
+    def __init__(self):
+        super().__init__(name="TensorFlow Record File", t=TFRecordsDirectory)
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/record.py:162]:
```suggestion
        super().__init__(name="TensorFlow Record Directory", t=TFRecordsDirectory)
```

File: flytekit/extras/tensorflow/record.py [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+        return Literal(scalar=Scalar(blob=Blob(metadata=meta, uri=remote_path)))
+
+    def to_python_value(
+        self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[TFRecordsDirectory]
+    ) -> TFRecordDatasetV2:
+
+        uri, metadata = extract_metadata_and_uri(lv, expected_python_type)
+        local_dir = ctx.file_access.get_random_local_directory()
+        files = os.scandir(uri)
+        filenames = [os.path.join(local_dir, f.name) for f in files]
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/record.py:176]: I think we need to get the file names from the local directory, not the remote path. In local case, it works, but when run on Flyte backend, it'll be a remote uri.

File: flytekit/extras/tensorflow/record.py [hunk 30]
```diff
@@ -0,0 +1,198 @@
...
+
+
+@overload
+def extract_metadata_and_uri(t: TFRecordFile) -> Tuple[TFRecordFile, TFRecordDatasetConfig]:
+    ...
+
+
+@overload
+def extract_metadata_and_uri(t: TFRecordsDirectory) -> Tuple[TFRecordsDirectory, TFRecordDatasetConfig]:
+    ...
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/record.py:46]: Just wondering why we need these...

File: flytekit/types/directory/__init__.py [hunk 31]
```diff
@@ -26,3 +27,11 @@
     This is usually the SummaryWriter output in PyTorch or Keras callbacks which record the history readable by
     TensorBoard.
 """
+
+tfrecords_dir = typing.TypeVar("tfrecords_dir")
```
Comment (samhita-alla) [inline flytekit/types/directory/__init__.py:31]: Can we say "tfrecord" since that's the file extension?

File: flytekit/types/file/__init__.py [hunk 32]
```diff
@@ -109,3 +109,8 @@ def check_and_convert_to_str(item: typing.Union[typing.Type, str]) -> str:
 #: Can be used to receive or return an ONNXFile. The underlying type is a FlyteFile type. This is just a
 #: decoration and useful for attaching content type information with the file and automatically documenting code.
 ONNXFile = FlyteFile[onnx]
+
+tfrecords_file = Annotated[str, FileExt("tfrecords_file")]
```
Comment (samhita-alla) [inline flytekit/types/file/__init__.py:113]: Can we say "tfrecord" since that's the file extension?

File: tests/flytekit/unit/extras/tensorflow/record/test_record.py [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+@task
+def t2(dataset: TFRecordFile):
+
+    # if not annotated with TFRecordDatasetConfig, all attributes should default to None
+    assert isinstance(dataset, TFRecordDatasetV2)
+    assert dataset._compression_type is None
+    assert dataset._buffer_size is None
+    assert dataset._num_parallel_reads is None
+
+
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/record/test_record.py:76]: Can we have a task to accept `TFRecordsDir` without the config as well?

**Comment by samhita-alla** [comment]
Amazing work, @ryankarlos! A few more comments. Sorry about incrementally reviewing the PR. :/

**Comment by ryankarlos** [comment]
> Amazing work, @ryankarlos! A few more comments. Sorry about incrementally reviewing the PR. :/

Thank you ! No thats fine, you have spotted a lot of my errors which is good !

**Review by samhita-alla — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
